### PR TITLE
feat(plugin-essentials): add --tree flag to yarn workspaces list

### PR DESCRIPTION
**Pull request by paul-soporan**
**What's the problem this PR addresses?**

`yarn workspaces list` doesn't currenly have an option to output the workspace tree instead of a list workspaces.

**How did you fix it?**

I added a new `-t,--tree` flag, that outputs the workspace tree instead of the list of workspaces.

A few implementation details:
- The workspace information tree is represented as a tuple instead of an object:
```ts
export type WorkspaceInfoTree = [{/* The information */}, Array<WorkspaceInfoTree>];
```
The reason for this is because, this way, we can always build the tree, and when the `--tree` flag is not present, we can simply flatten the tree into a list using `infoTree.flat(Infinity)`.
- Because of this, the current behavior (without the `--tree` flag) remains unchanged.
- When the `--tree` flag is used, we use a recursive IIFE to create a TreeObject that we can then pass to `treeify.asTree` to pretty-print the tree. Here I went with printing the pretty name instead of the relative path, because I feel it makes a bit more sense in the context of a tree, as relative paths would look weird unless resolved relative to their parent instead of the project root.
- When both the `--tree` and `--json` flags are used, the original `WorkspaceInfoTree` is outputted, as it wouldn't make sense to print an object, since keys can only be strings.

This is how the tree of the berry repo currently looks when outputted by the `yarn workspaces list -t` command:

image

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I have verified that all automated PR checks pass.

### CODE REVIEW

File: packages/yarnpkg-cli/sources/polyfills.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
+// Remove when dropping Node 10 (~April 2021)
+import 'array-flat-polyfill';
```
Comment (merceyz) [inline packages/yarnpkg-cli/sources/polyfills.ts:2]: We already have `lodash/flatten` in the bundle so you can use that instead and save 647 bytes ^^

**Comment by arcanis** [comment]
Is it really useful? Due to how it requires to use nested workspaces (which I expect to be very very rare, in most cases people will just list a single pattern such as `packages/*`, just like we do), the output still really looks like a list, perhaps with slightly fancier markers.

**Comment by paul-soporan** [comment]
Nested workspaces are the original use case I had in mind, because outputting them as a list doesn't look too good IMO. Even we have a nested workspaces case with the `acceptance-tests` package. That's why I felt that it would be useful to better see the worktree, rather than a flat list of workspaces.

~~Note that, theoretically, even a `packages/*` pattern is not a flat list, since it has a root workspace and child workspaces. :P~~

I still feel like this feature has its benefits and that it improves the output (personally, if this was implemented, I'd never execute the `yarn workspaces list` command without the `--tree` flag, but that's my subjective opinion). I also feel that with bigger and bigger monorepos, nested workspaces won't be such a niche use case in the future.

What do you think?

**Comment by paul-soporan** [comment]
Another use case I haven't initially considered: Without this implementation, there's currently no way to build the workspace tree just from the output of `yarn workspaces list --json --verbose`, since the CWDs aren't enough.

**Comment by arcanis** [comment]
I'm going to close this PR since it's not clear whether we should land it - note that there's also the new `yarn info` command which prints the dependencies. Perhaps it solves a similar use case?

**Comment by paul-soporan** [comment]
It's fine, I also intended to close this PR but forgot about it. The lack of a tree display doesn't bother me that much anymore and I think it's fine to not add the extra complexity.

> note that there's also the new yarn info command which prints the dependencies. Perhaps it solves a similar use case?

It's similar but not the same. `yarn info` prints a tree display of the workspace dependency tree, while this PR makes it so that `yarn workspaces list` prints a tree display of the workspace *definition* tree.

**Pull request closed without merging**